docs/staking.md

```diff
@@ -0,0 +1,258 @@
+# Staking Contract Documentation
+
+## Overview
+The `Staking` contract manages SAFE token deposits tied to validator addresses. It enables:
+
+- Staking SAFE tokens toward registered validators.
+- Delayed withdrawals via per-staker-per-validator queues.
+- Owner-governed configuration and validator registry changes with a timelock.
+- Recovery of unencumbered tokens accidentally sent to the contract.
+
+The contract is non-upgradeable and relies on `Ownable` for admin control. It does not implement rewards, slashing, or validator performance logic; it is a ledger for deposits and delayed withdrawals.
+
+## Architecture Summary
+- **Token**: Immutable `SAFE_TOKEN` ERC20.
+- **Stake accounting**: Per staker and per validator totals, plus global counters.
+- **Withdrawals**: Each staker/validator pair has a doubly linked list sorted by `claimableAt`.
+- **Config changes**: Two-step propose/execute with a fixed `CONFIG_TIME_DELAY`.
+- **Access control**: `onlyOwner` for proposals and token recovery; execution is permissionless once timelock expires.
+
+## Data Structures
+
+### Structs
+- `WithdrawalNode`
+  - **Purpose**: Node in a per-staker-per-validator withdrawal queue.
+  - **Fields**:
+    - `amount`: Withdrawal amount.
+    - `claimableAt`: Timestamp when claimable.
+    - `previous` / `next`: Linked list pointers (0 is null).
+  - **Invariants / edge cases**:
+    - Queue is expected to be sorted by `claimableAt` ascending.
+    - Linked list pointers must remain consistent; incorrect manual insertion reverts.
+    - `claimableAt` is stored as `uint128`; truncation only possible at extreme timestamps.
+
+- `WithdrawalQueue`
+  - **Purpose**: Tracks head/tail of a withdrawal list for a staker/validator pair.
+  - **Invariants**:
+    - `head == 0 && tail == 0` indicates empty queue.
```

Comment:
Does that mean that `head != 0 <=> tail != 0`? If so, I would add that as an invariant here as well.